### PR TITLE
Make `ConstantEvaluator` truncate the result of shift operations

### LINKED ISSUE #16596: `ConstantEvaluator` does not truncate the result of left shift operations
The `ConstantEvaluator` attempts to convert the result of a left shift operation to the target type, which fails if the value is out of range.
This happens, for example, when evaluating array lengths or storage layout specifiers. 

The behavior is inconsistent with codegen, where the value is cleaned up either lazily (evmasm) or at the end of the operation (IR) to fit the type width.

The [documentation section on shifts](https://docs.soliditylang.org/en/latest/types.html#shifts) describes the correct process, stating that the result of a shift operation has the type of the left operand, overflow checks are never performed and the value is always truncated.

This problem affects only left shifts, since for right shifts the result is always smaller than or equal to the left operand.
Also important to note that expressions formed only by literals have `RationalNumberType` and unlimited precision and are not susceptible to this error.

Minimal repro:
```solidity
// SPDX-License-Identifier: NONE
pragma solidity ^0.8.0;
uint8 constant FIVE = 5;
uint8 constant SIXTEEN = 16;
contract C {

    uint8 constant CONST_AND_LITERAL = SIXTEEN << 5;
    uint8 constant CONST_ONLY = SIXTEEN << FIVE;

    // TypeError 2643: Arithmetic error when computing constant value.
    uint[CONST_AND_LITERAL] array1;
    // TypeError 2643: Arithmetic error when computing constant value.
    uint[CONST_ONLY] array2;
    // TypeError 2643: Arithmetic error when computing constant value.
    uint[SIXTEEN << 5] array3;
    // TypeError 2643: Arithmetic error when computing constant value.
    uint[2**255 << ONE] array4;
    
    function f() public pure returns (uint8) {
        uint8 x = 16;
        // Ok, returns 0
        return x << 5;
    }
}
```

Analysis
---
After calculating the result of the binary operation, ConstantEvaluator tries to convert the value, which is held as an unlimited-precision rational number, to the type of the operation result.

The computed value of the operation should be truncated to the type of the left operand as stated in the [shift operator documentation section](https://docs.soliditylang.org/en/latest/types.html#shifts).
Note that in this case, literals are assigned either `int256` or `uint256`.
ConstantEvaluator instead of truncating, tries converting the value to the left operand type.
This triggers error `2643`, whenever the value does not fit into the type of the operation result.

Codegen however performs cleanup/masking of the result so it fits properly into the left operand type.

In the evmasm pipeline, the final value is not truncated immediately.
The compiler does not clean the bits if the following operation is not affected, it only performs the cleanup if the next operation uses the value.

On the other hand, the IR pipeline explicitly does the cleanup on the result (see `YulUtilFunctions::typedShiftLeftFunction()`).

Either way, the result is always truncated to the result type.

`ConstantEvaluator` needs to process shift operation results in a separate function/branch from ordinary (mostly arithmetic) operations.
I suggest that a `cleanup` function is added and used after shift operation is evaluated.

### PR DESCRIPTION
**Pull request by matheusaaguiar**
Fix #16596.
Spotted in https://github.com/argotorg/solidity/pull/16456#discussion_r2860213445.

### CODE REVIEW

File: libsolidity/analysis/ConstantEvaluator.cpp [hunk 1]
```diff
@@ -411,7 +434,9 @@ void ConstantEvaluator::endVisit(BinaryOperation const& _operation)
 		std::get<rational>(right.value)
 	))
 	{
-		TypedValue convertedValue = convertType(*value, *resultType);
+		TypedValue convertedValue = TokenTraits::isShiftOp(_operation.getOperator()) ?
```
Comment (matheusaaguiar) [inline libsolidity/analysis/ConstantEvaluator.cpp:437]: I guess it could be done only for left shift.

**Comment by matheusaaguiar** [comment]
Not sure about a Changelog entry...

File: libsolidity/analysis/ConstantEvaluator.cpp [hunk 2]
```diff
@@ -275,6 +275,30 @@ TypedValue convertType(TypedValue const& _value, Type const& _type)
...
+TypedValue cleanupValue(rational const& _value, Type const& _type)
+{
+	if (_type.category() != Type::Category::Integer)
+		return convertType(_value, _type);
+
+	auto const* integerType = dynamic_cast<IntegerType const*>(&_type);
+	solAssert(integerType);
+
+	solAssert(_value.denominator() == 1);
+	bigint integerValue = _value.numerator() / _value.denominator();
```
Comment (clonker) [inline libsolidity/analysis/ConstantEvaluator.cpp]:
```suggestion
	bigint integerValue = _value.numerator();
```

you assert that denom is 1 right above

File: libsolidity/analysis/ConstantEvaluator.cpp [hunk 3]
```diff
@@ -275,6 +275,30 @@ TypedValue convertType(TypedValue const& _value, Type const& _type)
...
 
+TypedValue cleanupValue(rational const& _value, Type const& _type)
+{
+	if (_type.category() != Type::Category::Integer)
+		return convertType(_value, _type);
+
+	auto const* integerType = dynamic_cast<IntegerType const*>(&_type);
+	solAssert(integerType);
+
+	solAssert(_value.denominator() == 1);
```
Comment (clonker) [inline libsolidity/analysis/ConstantEvaluator.cpp:288]: why can we assume that, because it's an `IntegerType`? Is `rational` always in reduced/normalized form?

File: libsolidity/analysis/ConstantEvaluator.cpp [hunk 4]
```diff
@@ -275,6 +275,30 @@ TypedValue convertType(TypedValue const& _value, Type const& _type)
...
+	bigint integerValue = _value.numerator() / _value.denominator();
+
+	unsigned int numBits = integerType->numBits();
+	bigint mask = (bigint(1) << numBits) - 1;
+	bigint sign = bigint(1) << (numBits - 1);
+	// clean bits out of range
+	integerValue = integerValue & mask;
+
+	// extend sign if needed
+	if (integerType->isSigned() && (integerValue & sign))
```
Comment (clonker) [inline libsolidity/analysis/ConstantEvaluator.cpp]:
```suggestion
    bool const isNegative = integerType->isSigned() && boost::multiprecision::bit_test(integerValue, numBits - 1);
	if (isNegative)
```

then the sign mask above isn't needed anymore

File: libsolidity/analysis/ConstantEvaluator.cpp [hunk 5]
```diff
@@ -411,7 +435,9 @@ void ConstantEvaluator::endVisit(BinaryOperation const& _operation)
 		std::get<rational>(right.value)
 	))
 	{
-		TypedValue convertedValue = convertType(*value, *resultType);
+		TypedValue convertedValue = TokenTraits::isShiftOp(_operation.getOperator()) ?
+			cleanupValue(*value, *resultType) :
+			convertType(*value, *resultType);
```
Comment (clonker) [inline libsolidity/analysis/ConstantEvaluator.cpp]: Personally I'd rather introduce a helper `rational truncateToType(rational const& _value, IntegerType const& _type)` and then do the conversion directly here. that would avoid mixing concerns where, right now, `cleanupValue` both converts and truncates. I imagine it could read something like


```suggestion
		rational resultValue = *value;
		if (isShiftOp)
			if (auto const* integerType = dynamic_cast...)
				resultValue = truncateToType(resultValue, *integerType);
		
		TypedValue convertedValue = convertType(resultValue, *resultType);
```

**Review by clonker — COMMENTED** [review]
Some code structure feedback and questions. Overall looks correct to me. I think `BitNot` has similar issues. For example `~uint8(0)` and `~uint8(-128)` both fail constant eval. If you agree, this could be folded into this pr or be a follow up.

> Not sure about a Changelog entry...

It's user-facing, isn't it? So imo having an entry is the right move.

File: libsolidity/analysis/ConstantEvaluator.cpp (hunk 5 above)
Comment (matheusaaguiar) [reply libsolidity/analysis/ConstantEvaluator.cpp]: Yeah, was concerned too about `cleanupValue` calling `convertType` at the end...will implement your suggestion.

File: libsolidity/analysis/ConstantEvaluator.cpp (hunk 3 above)
Comment (matheusaaguiar) [reply libsolidity/analysis/ConstantEvaluator.cpp:288]: Because it's `IntegerType`. IIRC, in comptime, it is always normalized. I will confirm this.

File: libsolidity/analysis/ConstantEvaluator.h [hunk 6]
```diff
@@ -38,19 +38,32 @@ namespace solidity::frontend
...
+ * Small drop-in replacement for TypeChecker to evaluate simple expressions of integer and string constants.
  *
  * Note: This always use "checked arithmetic" in the sense that any over- or underflow
  * results in "unknown" value.
  */
 class ConstantEvaluator: private ASTConstVisitor
 {
 public:
-	struct TypedValue
+	class TypedValue
```
Comment (matheusaaguiar) [inline libsolidity/analysis/ConstantEvaluator.h:50]: This refactor was suggested [here](https://github.com/argotorg/solidity/pull/15968#discussion_r2852861376) and it is simple enough that I think it can be slipped in this PR...

File: test/libsolidity/smtCheckerTests/operators/constant_evaluation_bitwise_not.sol [hunk 7]
```diff
@@ -7,3 +7,4 @@ contract C {
 // ====
 // SMTEngine: chc
 // ----
+// Warning 6031: (186-199): Internal error: Expression undefined for SMT solver.
```
Comment (matheusaaguiar) [inline test/libsolidity/smtCheckerTests/operators/constant_evaluation_bitwise_not.sol]: This is actually due to an issue in the SMT Checker.
I was worried that it had something to do with the changes in this PR and investigated it.
Apparently, the `SMTEncoder` visitor was written when only literals were constant evaluated.
Later when support for constant folding of variables was added, the relevant part of SMT Checker was not properly adapted and since there were no tests covering, it went unnoticed.

In short, when visiting the unary operation, SMT Encoder checks whether the result is a rational number, which would indicate a literal expression as operand.
Since literal expressions have unlimited precision, it would then skip visiting the operand.
However, the Constant Evaluator later gained the ability to evaluate other expressions and the SMT Encoder then skips visiting those, assuming they are literals.

For example, using unary minus can also trigger the warning:
```solidity
contract C {
    int256 constant signedConstant = 42;
    function test() public pure returns (int) {
        return -signedConstant;
    }
}
```

Before the changes in this PR, this test was not triggering the warning because the bitwise not was generating a value out of range of integer, and thus the Constant Evaluator was not returning a rational number, which would make that the operand was visited.

I will open an issue for this.

File: libsolidity/analysis/ConstantEvaluator.cpp (hunk 3 above)
Comment (clonker) [reply libsolidity/analysis/ConstantEvaluator.cpp:288]: and?

File: test/libsolidity/semanticTests/constantEvaluator/bit_not_operation_runtime_equivalence.sol [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    int8 constant NEGATIVE_SIGNED = ~I128; // = 127 (0111 1111)
+    uint[NEGATIVE_SIGNED] b;
+    int8 constant POSITIVE_SIGNED = ~IONE; // = -2 (1111 1110)
+    uint[POSITIVE_SIGNED * -1] c;
+    function testUnsignedEquivalence() public view returns (bool) {
+        uint8 runTimeResult = ~U253;
+
+            return
+                UNSIGNED == runTimeResult &&
+                a.length == runTimeResult;
```
Comment (clonker) [inline test/libsolidity/semanticTests/constantEvaluator/bit_not_operation_runtime_equivalence.sol]:
```suggestion
        return
            UNSIGNED == runTimeResult &&
            a.length == runTimeResult;
```

File: test/libsolidity/semanticTests/constantEvaluator/shift_signed_left_operand_runtime_equivalence.sol [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    uint[CONST_WRAP * -1] c;
+    int16 constant CONST_SIGN_CHANGED = I16_POSITIVE_127 << 9; // = -512 (1111 1110 0000 0000)
+    uint[CONST_SIGN_CHANGED * -1] d;
+
+    function testLiteralWrapEquivalence() public view returns (bool) {
+        int256 runTimeResult = -2**255 << ONE;
+
+            return
+                LITERAL_WRAP == runTimeResult &&
+                a.length == 1;
```
Comment (clonker) [inline test/libsolidity/semanticTests/constantEvaluator/shift_signed_left_operand_runtime_equivalence.sol]:
```suggestion
        return
            LITERAL_WRAP == runTimeResult &&
            a.length == 1;
```

File: Changelog.md [hunk 10]
```diff
@@ -29,6 +29,7 @@ Compiler Features:
 * Yul Optimizer: Improve performance of control flow side effects collector and function references resolver.
 
 Bugfixes:
+* Constant Evaluator: Fix incorrect result of bit operations which were not consistent with codegen.
```
Comment (clonker) [inline Changelog.md]:
```suggestion
* Constant Evaluator: Fix incorrect result of bitwise `not` and `shift` which were not consistent with codegen.
```

File: libsolidity/analysis/ConstantEvaluator.cpp [hunk 11]
```diff
@@ -470,3 +501,35 @@ void ConstantEvaluator::endVisit(FunctionCall const& _functionCall)
...
+		[&](rational const&) {
+			solAssert(dynamic_cast<RationalNumberType const*>(_type) || dynamic_cast<IntegerType const*>(_type));
+		},
+		[&](std::monostate const&) {
+			solAssert(!_type);
+		}
+	}, _value);
+
+	m_type = _type;
+	m_value = _value;
```
Comment (clonker) [inline libsolidity/analysis/ConstantEvaluator.cpp]: the `_value` can contain a `string` and this copies it around needlessly. i suggest `std::move`ing this into `m_value`.

```suggestion
	m_value = std::move(_value);
```

To make this work, you'd also have to drop the `const` in the constructor, ie, `TypedValue::TypedValue(Type const* _type, TypedValue::Value _value)`, otherwise `_value` is not movable.

File: test/libsolidity/semanticTests/constantEvaluator/bit_not_operation_runtime_equivalence.sol
Comment (clonker) [inline test/libsolidity/semanticTests/constantEvaluator/bit_not_operation_runtime_equivalence.sol:1]: You're testing
- `~uint8(253) == 2`
- `~int8(-128) == 127`
- `~int8(1) == 2`

I'd add the following corner cases to tie it down:
- on full 256 bit width tie down the boundaries
  - `~uint256(0) == 2**256 - 1`, `~uint256(uint256.max) == 0`
  - signed boundaries: `~int256.min` producing `max` and vice versa
- analogously complete the `uint8` and `int8` min/max boundaries

**Review by clonker — COMMENTED** [review]
looks pretty good to me! i'd love to see these additional tests (wouldn't expect anything to fail there but at least then the boundaries are tied down) and the unnecessary string copying gone :)

File: libsolidity/analysis/ConstantEvaluator.cpp (hunk 3 above)
Comment (matheusaaguiar) [reply libsolidity/analysis/ConstantEvaluator.cpp:288]: I confirmed and boost rational [docs](https://www.boost.org/doc/libs/latest/libs/rational/rational.html#Class%20rational%20synopsis) say they are normalized:
`rational numbers are exact as long as the numerator and denominator (which are always held in normalized form, with no common factors) are within the range of the underlying integer type.`

File: libsolidity/analysis/ConstantEvaluator.cpp (hunk 11 above)
Comment (matheusaaguiar) [reply libsolidity/analysis/ConstantEvaluator.cpp]: Done.

File: test/libsolidity/semanticTests/constantEvaluator/bit_not_operation_runtime_equivalence.sol
Comment (matheusaaguiar) [reply test/libsolidity/semanticTests/constantEvaluator/bit_not_operation_runtime_equivalence.sol:1]: Separated signed and unsigned cases in their own files and added the suggested cases.

File: libsolidity/analysis/ConstantEvaluator.cpp (hunk 3 above)
Comment (clonker) [reply libsolidity/analysis/ConstantEvaluator.cpp:288]: nice!

File: libsolidity/analysis/ConstantEvaluator.cpp [hunk 12]
```diff
@@ -390,29 +417,33 @@ void ConstantEvaluator::endVisit(BinaryOperation const& _operation)
...
 	))
 	{
-		TypedValue convertedValue = convertType(*value, *resultType);
-		if (!convertedValue.type)
+		rational resultValue = *value;
+		if (TokenTraits::isShiftOp(_operation.getOperator()))
+			resultValue = truncateToType(*value, *resultType);
+
+		TypedValue convertedValue = convertType(resultValue, *resultType);
+			if (!convertedValue.type())
```
Comment (nikola-matic) [inline libsolidity/analysis/ConstantEvaluator.cpp]:
```suggestion
		if (!convertedValue.type())
```

File: libsolidity/analysis/ConstantEvaluator.h [hunk 13]
```diff
@@ -38,19 +38,32 @@ namespace solidity::frontend
...
 	{
+	public:
+		using Value = std::variant<std::monostate, rational, std::string>;
+
+		TypedValue(): m_type(nullptr), m_value(std::monostate()) {}
+		TypedValue(Type const* _type, Value const _value);
```
Comment (nikola-matic) [inline libsolidity/analysis/ConstantEvaluator.h]:
```suggestion
		TypedValue(Type const* _type, Value  _value);
```
 Implementation doesn't designate `_value` as `const` from what I can tell, and since you're moving it, I assume this a leftover.

File: test/libsolidity/semanticTests/constantEvaluator/bit_not_signed_runtime_equivalence.sol [hunk 14]
```diff
@@ -0,0 +1,64 @@
+
```
Comment (nikola-matic) [inline test/libsolidity/semanticTests/constantEvaluator/bit_not_signed_runtime_equivalence.sol]:
```suggestion
```

File: test/libsolidity/semanticTests/constantEvaluator/bit_not_unsigned_runtime_equivalence.sol [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+            NOT_U8_ZERO == notZeroRuntimeResult &&
+            b.length == 255;
+    }
+    function testUnsignedEquivalence() public view returns (bool) {
+        uint[UNSIGNED] memory a;
+        uint8 runTimeResult = ~U253;
+
+            return
+                UNSIGNED == runTimeResult &&
+                a.length == 2;
```
Comment (nikola-matic) [inline test/libsolidity/semanticTests/constantEvaluator/bit_not_unsigned_runtime_equivalence.sol]: shift+tab

File: test/libsolidity/semanticTests/constantEvaluator/shift_signed_left_operand_runtime_equivalence.sol [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+            a.length == 1;
+    }
+
+    function testConstNoWrapEquivalence() public view returns (bool) {
+        int8 runTimeResult = I8_NEGATIVE_63 << 1;
+
+        return
+                CONST_NO_WRAP == runTimeResult &&
+                b.length == 126;
+    }
```
Comment (nikola-matic) [inline test/libsolidity/semanticTests/constantEvaluator/shift_signed_left_operand_runtime_equivalence.sol]: shift+tab

File: test/libsolidity/semanticTests/constantEvaluator/shift_signed_left_operand_runtime_equivalence.sol [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+
+    function testConstWrapEquivalence() public view returns (bool) {
+        int8 runTimeResult = I8_POSITIVE_127 << 1;
+
+        return
+                CONST_WRAP == runTimeResult &&
+                c.length == 2;
```
Comment (nikola-matic) [inline test/libsolidity/semanticTests/constantEvaluator/shift_signed_left_operand_runtime_equivalence.sol]: shift+tab

File: libsolidity/analysis/ConstantEvaluator.cpp [hunk 18]
```diff
@@ -390,29 +417,33 @@ void ConstantEvaluator::endVisit(BinaryOperation const& _operation)
...
-		std::get<rational>(left.value),
-		std::get<rational>(right.value)
+		left.asRational(),
+		right.asRational()
 	))
 	{
-		TypedValue convertedValue = convertType(*value, *resultType);
-		if (!convertedValue.type)
+		rational resultValue = *value;
+		if (TokenTraits::isShiftOp(_operation.getOperator()))
```
Comment (nikola-matic) [inline libsolidity/analysis/ConstantEvaluator.cpp]: Wouldn't right shifts go through this path as well? I'd add a couple of tests just in case.

File: libsolidity/analysis/ConstantEvaluator.h (hunk 13 above)
Comment (matheusaaguiar) [reply libsolidity/analysis/ConstantEvaluator.h]: Yeah, true. I forgot to remove this after clonker's suggestion.

File: libsolidity/analysis/ConstantEvaluator.cpp (hunk 18 above)
Comment (matheusaaguiar) [reply libsolidity/analysis/ConstantEvaluator.cpp]: Yes, it will. Although I [think it is not necessary](https://github.com/argotorg/solidity/pull/16597#discussion_r3084560298), because right shift result will always decrease in magnitude and need less bit width...
I will double-check that it's ok to skip this for right shifts.

Comment (matheusaaguiar) [reply libsolidity/analysis/ConstantEvaluator.cpp]: But I guess it is good to add some tests for consistency.

File: test/libsolidity/smtCheckerTests/operators/constant_evaluation_bitwise_not.sol (hunk 7 above)
Comment (clonker) [reply test/libsolidity/smtCheckerTests/operators/constant_evaluation_bitwise_not.sol]: The test has gone from the PR, is there an issue now?

File: libsolidity/analysis/ConstantEvaluator.h [hunk 19]
```diff
@@ -38,19 +38,32 @@ namespace solidity::frontend
 class TypeChecker;
 
 /**
- * Small drop-in replacement for TypeChecker to evaluate simple expressions of integer constants.
+ * Small drop-in replacement for TypeChecker to evaluate simple expressions of integer and string constants.
  *
  * Note: This always use "checked arithmetic" in the sense that any over- or underflow
  * results in "unknown" value.
```
Comment (clonker) [inline libsolidity/analysis/ConstantEvaluator.h:44]: `~` and `<<` wrap now, don't they?

File: test/libsolidity/semanticTests/constantEvaluator/shift_right_unsigned_runtime_equivalence.sol [hunk 20]
```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+uint256 constant TWO = 1;
```
Comment (clonker) [inline test/libsolidity/semanticTests/constantEvaluator/shift_right_unsigned_runtime_equivalence.sol]: 2 = 1? :D

File: test/libsolidity/semanticTests/constantEvaluator/shift_right_unsigned_runtime_equivalence.sol [hunk 21]
```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
```
Comment (clonker) [inline test/libsolidity/semanticTests/constantEvaluator/shift_right_unsigned_runtime_equivalence.sol]:
```suggestion
```

File: test/libsolidity/semanticTests/constantEvaluator/shift_right_signed_runtime_equivalence.sol [hunk 22]
```diff
@@ -0,0 +1,54 @@
+uint256 constant U_256 = 256;
+int8 constant I8_NEGATIVE_63 = -63;
+int8 constant I8_POSITIVE_127 = 127;
+int8 constant I8_NEGATIVE_128 = -128;
+int16 constant I16_POSITIVE_127 = 127;
```
Comment (clonker) [inline test/libsolidity/semanticTests/constantEvaluator/shift_right_signed_runtime_equivalence.sol]:
```suggestion
```

unused

File: test/libsolidity/semanticTests/constantEvaluator/bit_not_signed_runtime_equivalence.sol [hunk 23]
```diff
@@ -0,0 +1,63 @@
+int8 constant I8_NEGATIVE_128 = -128; // 1000 0000
+int8 constant I8_POSITIVE_127 = 127; // 0111 1111
```
Comment (clonker) [inline test/libsolidity/semanticTests/constantEvaluator/bit_not_signed_runtime_equivalence.sol]:
```suggestion
```

unused

File: test/libsolidity/semanticTests/constantEvaluator/shift_right_signed_runtime_equivalence.sol
Comment (clonker) [inline test/libsolidity/semanticTests/constantEvaluator/shift_right_signed_runtime_equivalence.sol:1]: in shift right no wrapping should occur, so i find `_WRAP` a bit misleading in the variable names

File: test/libsolidity/semanticTests/constantEvaluator/shift_right_unsigned_runtime_equivalence.sol
Comment (clonker) [inline test/libsolidity/semanticTests/constantEvaluator/shift_right_unsigned_runtime_equivalence.sol:1]: in shift right no wrapping should occur, so i find `_WRAP` a bit misleading in the variable names

**Review by clonker — COMMENTED** [review]
Hey, sorry for not getting back to this any sooner.

There is no test covering left shift by >= type width with a small type (e.g. `U8_64 << 8`); the wrap tests all shift by less than the width except the full-256-bit case, right? If so, I think it should be added.

There's also another small inconsistency with how the runtime behaves wrt large shift constants:

```solidity
uint256 constant ONE = 1;
uint256 constant A = ONE << 300;   // exact 2^300 fits the 4096-bit precision cap
uint256 constant B = ONE << 5000;  // exact 2^5000 exceeds it

contract C {
    uint[A + 1] a;                                        // needs compile-time evaluation
    uint[B + 1] b;                                        // needs compile-time evaluation
    function f() public pure returns (uint256) { return B; }  // runtime only
}
```

the `a` case compiles under this PR, the `b` case doesn't. It's probably not extremely important but still. :)

File: test/libsolidity/smtCheckerTests/operators/constant_evaluation_bitwise_not.sol (hunk 7 above)
Comment (matheusaaguiar) [reply test/libsolidity/smtCheckerTests/operators/constant_evaluation_bitwise_not.sol]: It was appearing in the PR only because of the issue I mentioned in the comment and #16654 fixed it.

File: test/libsolidity/semanticTests/constantEvaluator/shift_right_unsigned_runtime_equivalence.sol (hunk 20 above)
Comment (matheusaaguiar) [reply test/libsolidity/semanticTests/constantEvaluator/shift_right_unsigned_runtime_equivalence.sol]: :laughing: 
Fixed.

ps: But there is proof, I saw it in the interwebs.

File: libsolidity/analysis/ConstantEvaluator.h (hunk 19 above)
Comment (matheusaaguiar) [reply libsolidity/analysis/ConstantEvaluator.h:44]: Added a line explaining "~" and "<<" are the exception.

**Comment by matheusaaguiar** [comment]
> the `a` case compiles under this PR, the `b` case doesn't. It's probably not extremely important but still. :)

Yes, in `b` case the constant evaluator returns "empty" or no value because it overflows the 4096 bit width, however codegen executes outputting `0` as a result.
I modified the code so constant evaluator returns 0 as well. I am running some tests to check there's no problem.

File: libsolidity/analysis/ConstantEvaluator.cpp [hunk 24]
```diff
@@ -274,7 +274,30 @@ TypedValue convertType(TypedValue const& _value, Type const& _type)
 		[&](std::monostate const&) {
 			return TypedValue{};
 		}
-	}, _value.value);
+	}, _value.value());
+}
+
+rational truncateToType(rational const& _value, Type const& _type)
+{
+	if (_type.category() == Type::Category::RationalNumber)
```
Comment (matheusaaguiar) [inline libsolidity/analysis/ConstantEvaluator.cpp:282]: I changed this now, before it was `!= Type::Category::Integer`, because shifts work for FixedBytes type too, so this would return the value unaltered...
The compiler does not have constants of such type yet, but they are on the horizon.

File: libsolidity/analysis/ConstantEvaluator.cpp [hunk 25]
```diff
@@ -188,7 +188,7 @@ std::optional<rational> ConstantEvaluator::evaluateBinaryOperator(Token _operato
 		else if (_right < 0)
 			return std::nullopt;
 		else if (_right > std::numeric_limits<uint32_t>::max())
-			return std::nullopt;
+			return 0;
```
Comment (matheusaaguiar) [inline libsolidity/analysis/ConstantEvaluator.cpp:191]: Originally this PR didn't intend to touch right shifts, but this part needed adjustment to make the operation consistent with codegen.

File: test/libsolidity/syntaxTests/nameAndTypeResolution/370_shift_constant_left_excessive_rvalue_truncated.sol [hunk 26]
```diff
@@ -0,0 +1,4 @@
+contract C {
```
Comment (matheusaaguiar) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/370_shift_constant_left_excessive_rvalue_truncated.sol:1]: I renamed this and the next test file adding the suffix `truncated` because before the constant evaluator would be used to compute the value of the literal expressions and return an error due to the exp being greater than `uint32.max`.
However, codegen just truncates and accepts the operation.

**Review by matheusaaguiar — COMMENTED** [review]
@clonker I addressed the problem you found and made a few changes.

File: libsolidity/analysis/ConstantEvaluator.h (hunk 6 above)
Comment (blishko) [reply libsolidity/analysis/ConstantEvaluator.h]: I actually think this PR could be reviewed more easily if the refactor was separated out to its own PR.

**Comment by john23armedilla-spec** [comment]
Were can ifound my acount?

**Comment by matheusaaguiar** [comment]
@blishko I removed the refactor as you suggested in the solidity-dev channel.

File: libsolidity/analysis/ConstantEvaluator.cpp (hunk 25 above)
Comment (nikola-matic) [reply libsolidity/analysis/ConstantEvaluator.cpp:191]: Wait, doesn't SAR interpret this as two's complement, i.e. signed? If the left is negative, shouldn't this yield -1 instead of zero? Can you cover this with tests, e.g. `-1 >> 0x100000000`?

File: test/libsolidity/syntaxTests/types/rational_number_bitshift_limit.sol [hunk 27]
```diff
@@ -2,12 +2,9 @@ contract c {
     function f() public pure {
         int a;
         a = 1 << 4095; // shift is fine, but result too large
-        a = 1 << 4096; // too large
-        a = (1E1233) << 2; // too large
+        a = 1 << 4096; // too large, result is truncated to 0
```
Comment (nikola-matic) [inline test/libsolidity/syntaxTests/types/rational_number_bitshift_limit.sol:5]: Honestly I don't know if I'm reading this correctly, but it seems like this should still be an error.

From the docs:
[Number literal expressions retain arbitrary precision until they are converted to a non-literal type (i.e. by using them together with anything other than a number literal expression (like boolean literals) or by explicit conversion). This means that computations do not overflow and divisions do not truncate in number literal expressions.](https://docs.soliditylang.org/en/latest/types.html#rational-and-integer-literals)

This reads like the actual binary operation with two literals has an undecided type until the operation is computed, but this is clearly impossible here? On the other hand, I don't see why we should treat literals and typed values differently. Maybe @cameel can chime in?

Comment (cameel) [reply test/libsolidity/syntaxTests/types/rational_number_bitshift_limit.sol:5]: > This reads like the actual binary operation with two literals has an undecided type until the operation is computed, but this is clearly impossible here?

`1` and `4096` are both rational number literals. Operations on rational literals are performed in unlimited precision. This part of the expression is valid and I would expect it to produce the same value as `2**4096`.

On the other hand the assignment should not compile because it forces a conversion to a concrete type and the result is too large for `int`. If that's what you mean by "impossible" then you're right.

I haven't looked at the rest of PR but if this test passes then our `<<` in unlimited precision is broken. It should not truncate because the truncation is the artifact of assuming a specific precision. We should consider changing that in a breaking version, though that will be annoying because people might rely on current behavior. But I would expect that Core Solidity will ship with a correct version anyway so maybe it's still worth changing to make the transition less jarring.

It should not affect this PR though, which was meant to address only limited precision operations.

> I don't see why we should treat literals and typed values differently.

Because we have different types and you can't match all of them. The best you can do is match `uint256` but that's somewhat arbitrary. If you do that then you'll have cases of different behavior in `uint248`, `uint240` and all the smaller types. The choice consistent with arbitrary types is not to truncate until the value is actually converted to a specific type.

This problem is not unique to left shift, though it's the only case where the discrepancy is observable. All the arithmetic operators would do this if not for the fact that we have errors on overflow in limited precision. `<<` is an odd one because the overflow behavior is expected and well understood by users so it's not reasonable to disallow it unlike in e.g. addition.

File: libsolidity/analysis/ConstantEvaluator.cpp (hunk 25 above)
Comment (cameel) [reply libsolidity/analysis/ConstantEvaluator.cpp:191]: > Originally this PR didn't intend to touch right shifts, but this part needed adjustment to make the operation consistent with codegen.

You should mention high-level changes like this in the issue. This is making the change broader and we should consider if that's really the direction we want to go.

In this case in particular I think we should pass the change by solcore to make sure they're implementing the same behavior for limited vs unlimited precision. Do their `<<` and `>>` work like what we're aiming for here?

File: test/libsolidity/syntaxTests/types/rational_number_bitshift_limit.sol (hunk 27 above)
Comment (nikola-matic) [reply test/libsolidity/syntaxTests/types/rational_number_bitshift_limit.sol:5]: > On the other hand the assignment should not compile because it forces a conversion to a concrete type and the result is too large for int. If that's what you mean by "impossible" then you're right.

Yes

File: libsolidity/analysis/ConstantEvaluator.cpp (hunk 25 above)
Comment (matheusaaguiar) [reply libsolidity/analysis/ConstantEvaluator.cpp]: > You should mention high-level changes like this in the issue. This is making the change broader and we should consider if that's really the direction we want to go.

Yeah, I will separate this into its own issue/PR.

> In this case in particular I think we should pass the change by solcore to make sure they're implementing the same behavior for limited vs unlimited precision. Do their << and >> work like what we're aiming for here?

Will check with them.

Comment (cameel) [reply libsolidity/analysis/ConstantEvaluator.cpp]: I think it's fine to have both in the same issue/PR. It's just a change of scope from very specific "make left shift truncate" to the broader "make comptime and runtime shift behavior consistent".

I just wanted to push back against quietly making a scope shift within a PR. Tracking everything that happens in PRs takes a lot of attention and should not be necessary to find out about something like this. High-level decisions of this kind should be bubbled up to the issue.

File: test/libsolidity/syntaxTests/types/rational_number_bitshift_limit.sol (hunk 27 above)
Comment (matheusaaguiar) [reply test/libsolidity/syntaxTests/types/rational_number_bitshift_limit.sol]: So, I refreshed my mind about this PR and review it from the start.
Operations with rational literals are performed with "unlimited precision", although in reality there is a hard cap of 4096 bit-width for the result.
Thus, I think that they should not be truncated and when exceeding the hard cap, the result cannot be computed, hence `std::nullopt`.

Operations with integer values however are subject to truncation to fit the type's width.
From the [docs](https://docs.soliditylang.org/en/v0.8.36/types.html#shifts): `The result of a shift operation has the type of the left operand, truncating the result to match the type.`
Additionally, when the left operand is a rational literal and the right operand is integer, the type of the result is either `uint256` or `int256`, so I am assuming the same rule for these cases.

I added a flag to `ConstantEvaluator::evaluateBinaryOperator` to indicate when truncating is expected for a shift operation, so the behavior for literals is kept and only results with integer type are truncated.
